### PR TITLE
ci(server): enhance performance 

### PR DESCRIPTION
**Pull request by yk-eukarya**
## Summary by CodeRabbit

- **New Features**
	- Introduced a new Continuous Integration (CI) workflow to streamline the build process for web, server, and worker components.
	- Added structured checks for pull requests, including title validation and labeling.

- **Bug Fixes**
	- Improved job dependencies and execution conditions for enhanced reliability in CI processes.

- **Documentation**
	- Updated workflow names and descriptions for clarity.

- **Refactor**
	- Restructured CI workflows to utilize `workflow_call` for better modularity and integration.
	- Enhanced caching strategy for yarn dependencies in the web component workflow.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    outputs:
+      web: ${{ steps.web.outputs.any_changed }}
+      server: ${{ steps.server.outputs.any_changed }}
+      worker: ${{ steps.worker.outputs.any_changed }}
+    steps:
+      - name: checkout
+        uses: actions/checkout@v4
+      - name: changed files for web
+        id: web
+        uses: tj-actions/changed-files@v36
```
Comment (KeisukeYamashita) [inline .github/workflows/ci.yml:18]:
```suggestion
        uses: tj-actions/changed-files@v45
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        with:
+          files: |
+            web
+            .github/workflows/ci.yml
+            .github/workflows/ci_web.yml
+
+      - name: changed files for server
+        id: server
+        uses: tj-actions/changed-files@v36
```
Comment (KeisukeYamashita) [inline .github/workflows/ci.yml:27]:
```suggestion
        uses: tj-actions/changed-files@v45
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        with:
+          files: |
+            server
+            .github/workflows/ci.yml
+            .github/workflows/ci_server.yml
+            .github/workflows/build_server.yml
+
+      - name: changed files for worker
+        id: worker
+        uses: tj-actions/changed-files@v36
```
Comment (KeisukeYamashita) [inline .github/workflows/ci.yml:37]:
```suggestion
        uses: tj-actions/changed-files@v45
```

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  push:
+    branches: [main, release]
+  pull_request:
+jobs:
+  prepare:
+    runs-on: ubuntu-latest
+    outputs:
+      web: ${{ steps.web.outputs.any_changed }}
+      server: ${{ steps.server.outputs.any_changed }}
+      worker: ${{ steps.worker.outputs.any_changed }}
```
Comment (KeisukeYamashita) [inline .github/workflows/ci.yml:12]:
```suggestion
      web: ${{ steps.web.outputs.any_modified }}
      server: ${{ steps.server.outputs.any_modified }}
      worker: ${{ steps.worker.outputs.any_modified }}
```

`any_changed` will not detect deletions, which we should.

Ref:  [tj-actions/changed-files](https://github.com/tj-actions/changed-files)

image

File: .github/workflows/ci_web.yml [hunk 5]
```diff
@@ -25,20 +9,12 @@ jobs:
...
-        id: yarn-cache-dir-path
-        run: echo "::set-output name=dir::$(yarn cache dir)"
-      - uses: actions/cache@v4
-        id: yarn-cache # use this to check for `cache-hit` (`steps.yarn-cache.outputs.cache-hit != 'true'`)
-        with:
-          path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
-          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
-          restore-keys: |
-            ${{ runner.os }}-yarn-
+          cache: yarn
```
Comment (KeisukeYamashita) [inline .github/workflows/ci_web.yml:16]: Question ✋ 

Does this cache strategy work?

`actions/setup-go` (the only action I know) stores the caches **per branch**,  if this `setup-node` action has the same strategy (likely be same), then it won't work as expected.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    if: needs.prepare.outputs.server == 'true'
+    uses: ./.github/workflows/ci_server.yml
+
+  ci-worker:
+    name: worker
+    needs: prepare
+    if: needs.prepare.outputs.worker == 'true'
+    uses: ./.github/workflows/ci_worker.yml
+
+  ci:
```
Comment (KeisukeYamashita) [inline .github/workflows/ci.yml:63]: Question ✋ 

May I ask what is this for?

**Review by KeisukeYamashita — CHANGES_REQUESTED** [review]

File: .github/workflows/ci_web.yml (hunk 5 above)
Comment (yk-eukarya) [reply .github/workflows/ci_web.yml:16]: AFAIK the default branch cache is shared with other branches, and another useful use case is when multiple pushes to the same branch happen then the cache can be useful

image

File: .github/workflows/ci.yml (hunk 6 above)
Comment (yk-eukarya) [reply .github/workflows/ci.yml:63]: this is the required action to allow the merge

**Review by KeisukeYamashita — APPROVED** [review]
Thanks for the updates and your explanations!

File: .github/workflows/ci_server.yml [hunk 7]
```diff
@@ -1,64 +1,65 @@
...
-      - server/**
-      - .github/workflows/ci_server.yml
-      - .github/workflows/server_build.yml
-      - .github/workflows/deploy_test.yml
-      - .github/workflows/pr_title.yml
-      - .github/workflows/stage.yml
-      - .github/workflows/release.yml
+  workflow_call:
+env:
+  GO_VERSION: "1.23"
```
[the remark this answers is not in the document]
Comment (yk-eukarya) [reply .github/workflows/ci_server.yml:5]: Golang version 1.23 is already released!

**Pull request merged**